TNSAIonMatterInteractor.py

```diff
@@ -129,3 +97,65 @@ def saveData(self):
 
         series.flush()
         del series
+
+    def backengine(self):
+        pass
+
+    def run(self):
+        [mom, weight]= self.readSDF()
+        energy = np.square(mom)/(2*mp*e)
+        de=self.parameters.energy_bin
+        nb=math.floor(np.max(energy)/de)
+        counts = [0 for i in range(nb+1)]
+        binedges = [(i+1)*de for i in range(nb+1)]
+        print("Number of energy bins: %s" % len(binedges))
+        for en in energy:
+            index=math.floor(en/de)
+            counts[index] += weight[index]
+        [E, xs] = self.read_xsec()
+        xs = np.interp(binedges, E * 1.e6, xs * 1.e-28)
+
+        vx=[0]
+        for i in range(len(binedges)):
+            px = np.sqrt(2 * mp * binedges[i] * e)
+            Nd = counts[i]* self.parameters.ibeam_radius**2 / self.parameters.neutron_weight
+            inc = int(round(Nd * self.parameters.target_density * self.parameters.target_length * xs[i]))
+            if inc>0:
+                vx=np.append(vx, np.ones(inc)*px/mp)
+                self.Nn += inc
+
+        print("Number of neutron macroparticles:", self.Nn)
+        vx=vx[1:]
+
+        self.data = np.zeros(shape=(self.__dims, self.Nn))
+        vn = math.sqrt(2 * 2.45e6 * e / mp)
+        self.data[0] = 1.e6 * self.parameters.target_length * random(self.Nn)
+        r = 1.e6 * self.parameters.ibeam_radius * random(self.Nn)  # positions will be saved in units of micron
```

Comment:
This will result in a larger density of neutron origins in the center, as the distance between two points with low radius is lower than the distance between two points with large radius, and there is the same likelihood for both. If you want equal density over a circle, use sqrt(random(self.Nm)).

Reply:
Yes, it is correct.  I made the changes.